AWS DynamoDB Grain State Provider

galvesribeiro opened the pull request:
Implementation of #2006 which is part of #2005

gabikliot commented:
I think there is duplication with the other pr.

galvesribeiro commented:
@gabikliot actually it is not duplicated... I made a mistake... I should wait this PR get merged before create a new branch for the other one. What I did was create a branch locally from this one so I can use the DynamoDB code into the Membership provider :(

If you can merge this one I can fix the other one.

Thanks

gabikliot commented:
I can review the membership provider. Someone else can review the storage. For me to review the membership provider, please make its pr clean and only include MBR provider code. I already posted some comments there.

galvesribeiro commented:
@gabikliot yea I'll make it clear but like I said, this PR need to be merged first since I need `DynamoDBDataManager<T>` to use it inside MBR on `OrleansSiloInstanceManager`.

Lets wait someone merge it here so I can work on the MBR one. Thanks

veikkoeeva commented:
In case work continues, some short notes I believe are applicable to this too: https://gitter.im/dotnet/orleans?at=57a35c77836d2d021163d9a6.

gabikliot commented on src/OrleansAWSUtils/Storage/DynamoDBDataManager.cs:
```diff
@@ -0,0 +1,248 @@
+            return batch.ExecuteAsync();
+        }
+
+        public Task UpsertItemAsync(T entry)
```
We need a separate Update which always relates to the previous version and updates the new version if previous was not changed
AND
Insert,, which always inserts if there is no prev version or fails otherwise.
Then any higher level retries should be done on top of it.

We can still have upsert for some situations when we don't care, but 95% of the time we do care.
In general, I always say that upsert is very harmfull and most of the time people use it wrongly instead of properly: try insert, if not read and try update. Otherwise, we run into a danger of a blink write.

Specifically, in both storage provider and MBR provider you will need Update, tryInsert and you will not need upsert.

shayhatsor replied:
Adding to @gabikliot note. When developing a storage provider it's important to think of what happens on "duplicate activation". If two duplicate grains attempt a write to the db at the same time, one of them must fail. This is easily achieved by making sure all write operations send the expected version with the data. I think of it as a conditional upsert operation.

gabikliot commented on src/OrleansAWSUtils/Storage/DynamoDBDataManager.cs:
```diff
@@ -0,0 +1,248 @@
+
+        private DynamoDBContext ddbContext;
+        private AmazonDynamoDBClient ddbClient;
+        private DynamoDBOperationConfig operationConfig;
```
looks like this could be static, right?

gabikliot commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,334 @@
+
+            public Task Write(GrainStateRecord record)
+            {
+                return tableManager.UpsertItemAsync(record);
```
I think this is wrong. You need to call Update here.

gabikliot replied:
Look at https://github.com/dotnet/orleans/blob/master/src/OrleansAzureUtils/Providers/Storage/AzureTableStorage.cs#L496.
There is a good reason it does create OR update, but not upsert.
We want to fail if we try to do create but it already exists. Otherwise, we can overwrite by mistake.

gabikliot commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,334 @@
+            }
+            catch (Exception)
+            {
+                throw;
```
??

shayhatsor commented on src/Orleans.sln:
```diff
@@ -64,6 +64,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Nuget", "Nuget", "{014CD19F
 		NuGet\Microsoft.Orleans.Core.nuspec = NuGet\Microsoft.Orleans.Core.nuspec
 		NuGet\Microsoft.Orleans.CounterControl.nuspec = NuGet\Microsoft.Orleans.CounterControl.nuspec
 		NuGet\Microsoft.Orleans.EventSourcing.nuspec = NuGet\Microsoft.Orleans.EventSourcing.nuspec
+		Microsoft.Orleans.OrleansAWSUtils.nuspec = Microsoft.Orleans.OrleansAWSUtils.nuspec
```
all others are under `Nuget` folder

shayhatsor commented on src/OrleansAWSUtils/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,25 @@
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("OrleansAWSUtils")]
+[assembly: AssemblyDescription("Orleans - Windows Azure Helper Classes")]
```
fix here

galvesribeiro commented:
Guys, dont review the current code. I had changed it and will let you know whenever I get it reopened probably later today. Thanks

galvesribeiro commented:
![image](https://cloud.githubusercontent.com/assets/4714040/17449408/9787e1cc-5b30-11e6-8903-086d19c6891a.png)

Ok, @gabikliot @shayhatsor (and anyone who can) now it is ready for review. Remember that this issue is just the Grain State storage provider and the core DataManager that will be used on other components like MBR, Reminder, etc.

As soon as I got feedback I'll update the PR with the changes. 

I'll wait this PR get completed before I start the refactory on the MBR one.

Thanks!

galvesribeiro commented:
BTW, if anyone want to run the tests, just follow the instructions [here](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html) that explain how to quickly download and run DynamoDB. No install required. Just unzip and call the Java commandline as explined.

shayhatsor commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,415 @@
+    /// </summary>
+    public class DynamoDBStorage : IStorageProvider
+    {
+        internal const string TableNameDefaultValue = "OrleansGrainState";
```
a suggestion, move all the configuration parsing to `DynamoDBStorageConfig` in order to separate concerns

galvesribeiro replied:
OK

shayhatsor commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,415 @@
+        /// <see cref="IStorageProvider.ReadStateAsync"/>
+        public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            if (tableDataManager == null) throw new ArgumentException("GrainState-Table property not initialized");
```
I'm not sure this is needed, here and in other places. even then, it's not an `ArgumentException`

galvesribeiro replied:
OK

shayhatsor commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,415 @@
+
+        private string GetKeyString(GrainReference grainReference)
+        {
+            var key = string.Format("{0}_{1}", serviceId, grainReference.ToKeyString());
```
The implementation here and in similar places is the same as the Azure one. My knowledge of Azure Table or DynamoDB is close to nothing, so I don't know how similar they are. I think that before merging this, we must convince ourselves that following the Azure oriented implementation gives the best throughput in DynamoDB too.

galvesribeiro replied:
DDB is pretty much same thing then Azure tables, just in AWS.  Many of the rules on Azure apply to there and many parts of the code will remain the same which looks to me ok.

shayhatsor commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,415 @@
+
+            initMsg = String.Format("{0} UseJsonFormat={1}", initMsg, useJsonFormat);
+
+            Log.Info(ErrorCode.StorageProviderBase, "AWS DynamoDB Provider: {0}", initMsg);
```
I suggest you either use the Azure error codes (like in the previous version of this PR) or copy paste and rename them like we did in the ado.net storage provider PR. If you feel like fixing this the right way, the best would be to merge all the storage provider errors and put them in the Orleans providers project. see https://github.com/dotnet/orleans/pull/1682#issuecomment-236450895. the important thing is that you shouldn't use the base error because it'd make it harder later to refactor.

galvesribeiro replied:
That ErrorCode enum is something very weird to me. It contains error codes from Azure so I put this instead.

I think unless the errors are REALLY uniform across all providers (which should contain a set of common error codes) each provider should have its own error codes. 

I'll create a separated PR after all the providers for AWS are done and before we start creating Nugets, only enhancing the logging across all AWS providers.

jdom replied:
I agree you shouldn't use the base number

veikkoeeva replied:
In ADO.NET the base is moved higher than what it is in ATS or Blob storage (for reference [here](https://github.com/dotnet/orleans/blob/master/src/OrleansSQLUtils/Storage/Provider/AdoNetStorageProvider.cs#L28)). I tried to make it so they match the Azure ones, just with higher base.

shayhatsor commented on src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,415 @@
+                if (logger.IsVerbose3) logger.Verbose3(ErrorCode.StorageProviderBase, "Writing: PartitionKey={0} RowKey={1} to Table={2} with ETag={3}", record.PartitionKey, record.RowKey, TableName, record.ETag);
+                string eTag = record.ETag == 0 ?
+                    await tableManager.CreateTableEntryAsync(record).ConfigureAwait(false) :
+                    await tableManager.UpdateTableEntryAsync(record, record.ETag.ToString()).ConfigureAwait(false);
```
You have to wrap this with a try and catch and rethrow an `InconsistentStateException` when the write failed because the row already exists or there was a version conflict. Looking at the Azure code, this isn't implemented there, which is a bug.

a note for @veikkoeeva : having the insert and update separate will simplify the db scripts greatly (no need to lock anything). I don't know how I didn't think of it before.

galvesribeiro replied:
Done. The parse is only on DynamoDBStorage for all fields of the connection strings and not spread everywhere.

shayhatsor commented:
@galvesribeiro, about the tests you've added. How are they different than the tests already found in Orleans? Do they test specific DynamoDB issues ? I suggest making base classes for these tests. In general, the similarity between the Azure Table implementation and the DynamoDB seems like a good candidate for refactoring to base classes. What do you think?

galvesribeiro commented:
@shayhatsor The tests are most of them the same. Some I added, some I changed.

I already talked with @jdom while doing the xunit migration that we should have the black box tests migrated to theories so we test against the abstraction/interface surface with multiple provider implementation. But again, this is indeed a refactory we can do, but I think it is not the focus of this PR.

jdom commented:
We will not use theories, as each provider should be able to be developed independent from the rest of the solution, and even if it was, each of them requires a different fixture for set up. We will be moving to a test kit as explained in #2011. The more we add to the same Tester project, the harder it is to port to coreclr by partitioning the work. If this PR doesn't address the test kit refactoring, we can do that later anyway, as a prereq for the migration to coreclr

jdom commented on src/Orleans/Properties/AssemblyInfo.cs:
```diff
@@ -24,12 +24,12 @@
 [assembly: InternalsVisibleTo("OrleansRuntime")]
 [assembly: InternalsVisibleTo("OrleansHost")]
 [assembly: InternalsVisibleTo("OrleansAzureUtils")]
+[assembly: InternalsVisibleTo("OrleansAWSUtils")]
```
Why is this needed? In newer providers (such as the event hub stream provider) we started doing some work so that it doesn't require any internals, and any external provider could also be developed in the same way as built in ones. Is it possible to avoid this, and if not, let's try to find a way to make it better. 
The event hub provider did require some refactoring of the main Orleans assemblies, but in the end it improved them for all stream providers.

galvesribeiro replied:
Right now I just had to add that since I'm using ErrorCode enum. I'll create a new enum with all the AWS specific error codes once we finish the review of all the PRs so before we release the code thiss will be out.

galvesribeiro commented:
Ok. Reset on master and made some fixes. The other enhancements suggested here comes with other PR before we start to generate Nugets to the whole OrleansAWSUtils.

shayhatsor commented on src/OrleansAWSUtils/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,430 @@
+                var value = serviceConfig.Split(new[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
+                if (value.Length == 2 && !string.IsNullOrWhiteSpace(value[1]))
+                    service = value[1];
+            }
```
this block, including the line above it should be refactored as a method. now it's just duplicated 4 times

shayhatsor commented on src/OrleansAWSUtils/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,430 @@
+            }
+            catch (Exception)
+            {
+                throw;
```
?

shayhatsor commented on test/TestGrainInterfaces/IPersistenceTestGrains.cs:
```diff
@@ -19,7 +19,7 @@ public interface IPersistenceTestGrain : IGrainWithGuidKey
 
-    public interface IPersistenceTestGenericGrain<T> : IPersistenceTestGrain // IGrainWithGuidKey
+    public interface IPersistenceTestGenericGrain<T> : IPersistenceTestGrain // IGrainWithGuidKeyaws
     { }
```
?

galvesribeiro replied:
Ouch, let me revert that.

shayhatsor commented on src/OrleansAWSUtils/Storage/Provider/DynamoDBStorageProvider.cs:
```diff
@@ -0,0 +1,340 @@
+
+namespace Orleans.Storage
+{
+    public class DynamoDBStorageProvider : IStorageProvider
```
all public entities must have at least summary. Also, please add some design rationale, why DynamoDB, implementation details and specifics, why you chose certain defaults and what do they mean, etc.

galvesribeiro replied:
All provider implementations are never accessed directly so no-one will ever read comments. The only reason it is public is so it can be accessed by Orleans internals. I think is more useful a better documentation on the docs page.

shayhatsor replied:
When someone other than you tries to figure out your thought process when implementing this, they won't go to the Orleans documentation for details. Please add a short summary like I mentioned, I agree that a more inclusive and exhaustive documentation should be added to Orleans documentation. Also, having public entities without a summary raises compilation warnings in Orleans.

shayhatsor commented on src/Orleans/Properties/AssemblyInfo.cs:
```diff
@@ -24,12 +24,12 @@
 [assembly: InternalsVisibleTo("TestInternalGrainInterfaces")]
-[assembly: InternalsVisibleTo("TestInternalGrains")]
-
+[assembly: InternalsVisibleTo("TestInternalGrains")]
```
?

galvesribeiro replied:
When added `[assembly: InternalsVisibleTo("OrleansAWSUtils")]` just removed an empty line. Not that it worth...

shayhatsor commented on src/OrleansAWSUtils/AWSUtils.cs:
```diff
@@ -0,0 +1,66 @@
+    /// <summary>
+    /// Some basic utilities methods for AWS SDK
+    /// </summary>
+    public static class AWSUtils
```
can this be internal?

galvesribeiro replied:
Done!

shayhatsor commented on src/OrleansAWSUtils/Storage/DynamoDBStorage.cs:
```diff
@@ -0,0 +1,430 @@
+
+                    await Task.Delay(2000);
+
+                } while (description.TableStatus != TableStatus.ACTIVE);
```
can this go on forever ?

galvesribeiro replied:
Nop, it throw an exception and in 99.99% of the cases the loop never execute twice but, AWS docs say that we have to wait for active. I just changed it a bit to make the loop only run while `TableStatus.CREATING`.

shayhatsor commented:
@galvesribeiro, thanks for this great contribution !

galvesribeiro commented:
@shayhatsor Thanks for having time to review and for the feedback :)
